datafusion/physical-plan/src/aggregates/row_hash.rs

```diff
@@ -639,6 +686,53 @@ pub(crate) fn create_group_accumulator(
     }
 }
 
+/// Check if we can enable the blocked optimization for `GroupValues` and `GroupsAccumulator`s.
+/// The blocked optimization will be enabled when:
+///   - When `enable_aggregation_blocked_groups` is true(default to true)
+///   - It is not streaming aggregation(because blocked mode can't support Emit::first(exact n))
+///   - The spilling is disabled(still need to consider more to support it efficiently)
+///   - The accumulator is not empty(I am still not sure about logic in this case)
+///   - [`GroupValues::supports_blocked_groups`] and all [`GroupsAccumulator::supports_blocked_groups`] are true
+///
+/// [`GroupValues::supports_blocked_groups`]: crate::aggregates::group_values::GroupValues::supports_blocked_groups
+/// [`GroupsAccumulator::supports_blocked_groups`]: datafusion_expr::GroupsAccumulator::supports_blocked_groups
+///
+// TODO: support blocked optimization in streaming, spilling, and maybe empty accumulators case?
```

Comment:
yes, I think the goal should be to support blocked optimizations in all these cases (so we can remove the existing code)

I can file tickets to help organize this work (obviously you don't have to do so)